grpc/src/main/java/com/linecorp/armeria/server/grpc/DefaultUnframedGrpcErrorHandler.java

```diff
@@ -0,0 +1,276 @@
+import com.linecorp.armeria.common.HttpResponse;
+import com.linecorp.armeria.common.HttpStatus;
+import com.linecorp.armeria.common.MediaType;
+import com.linecorp.armeria.common.ResponseHeaders;
+import com.linecorp.armeria.common.annotation.Nullable;
+import com.linecorp.armeria.common.grpc.protocol.GrpcHeaderNames;
+import com.linecorp.armeria.common.logging.RequestLogAccess;
+import com.linecorp.armeria.common.logging.RequestLogProperty;
+import com.linecorp.armeria.common.util.Exceptions;
+import com.linecorp.armeria.internal.common.JacksonUtil;
+import com.linecorp.armeria.internal.common.util.TemporaryThreadLocals;
+import com.linecorp.armeria.server.ServiceRequestContext;
+
+import io.grpc.Status;
+import io.grpc.Status.Code;
+
+final class DefaultUnframedGrpcErrorHandler {
+
+    private static final Logger logger = LoggerFactory.getLogger(DefaultUnframedGrpcErrorHandler.class);
+
+    private static final MessageMarshaller ERROR_DETAILS_MARSHALLER =
+            MessageMarshaller.builder()
+                             .omittingInsignificantWhitespace(true)
+                             .register(RetryInfo.getDefaultInstance())
+                             .register(ErrorInfo.getDefaultInstance())
+                             .register(QuotaFailure.getDefaultInstance())
+                             .register(DebugInfo.getDefaultInstance())
+                             .register(PreconditionFailure.getDefaultInstance())
+                             .register(BadRequest.getDefaultInstance())
+                             .register(RequestInfo.getDefaultInstance())
+                             .register(ResourceInfo.getDefaultInstance())
+                             .register(Help.getDefaultInstance())
+                             .register(LocalizedMessage.getDefaultInstance())
+                             .build();
+
+    private static final ObjectMapper mapper = JacksonUtil.newDefaultObjectMapper();
+
+    /**
+     * Ensure that unframedGrpcStatusMappingFunction never returns null by falling back to the default.
+     */
+    private static UnframedGrpcStatusMappingFunction ofStatusMappingFunction(
+            UnframedGrpcStatusMappingFunction statusMappingFunction) {
+        return requireNonNull(statusMappingFunction, "statusMappingFunction")
+                .orElse(UnframedGrpcStatusMappingFunction.of());
+    }
+
+    static JsonNode convertErrorDetailToJsonNode(List<Any> details)
+            throws IOException {
+        final StringWriter jsonObjectWriter = new StringWriter();
+        try (JsonGenerator jsonGenerator = mapper.createGenerator(jsonObjectWriter)) {
+            jsonGenerator.writeStartArray();
+            for (Any detail : details) {
+                ERROR_DETAILS_MARSHALLER.writeValue(detail, jsonGenerator);
+            }
+            jsonGenerator.writeEndArray();
+            jsonGenerator.flush();
+            return mapper.readTree(jsonObjectWriter.toString());
```

Comment:
It looks inefficient to encode protobuf to string and decode it to `JsonNode` again.
How about building the entire JSON result using `JsonGenerator`?
For example:
```java
final ByteBuf buffer = ctx.alloc().buffer();
try (ByteBufOutputStream outputStream = new ByteBufOutputStream(buffer);
     JsonGenerator jsonGenerator = mapper.createGenerator(outputStream)) {
    // write code, status, message, stack-trace and details with jsonGenerator
}
```
If you don't have any experiences handling `ByteBuf` with `HttpData`, please refer to:
https://github.com/line/armeria/blob/3dd45df368684a401ef5d9a4dba1166b5d4ee905/core/src/main/java/com/linecorp/armeria/server/metric/PrometheusExpositionService.java#L93-L102